bug: kuke attach fails with "connect: invalid argument" on long runPaths (SUN_PATH overflow)

## Context

`make e2e TestKuke_AttachDetach_KeepsTaskRunning` fails on `origin/main` (verified at 6bd245b) with:

```
Error: error attaching client: ping failed: dial unix /tmp/TestKuke_AttachDetach_KeepsTaskRunning2015614058/001/data/e-r-86/e-sp-52/e-st-4f/e-ce-c4/work/tty/socket: connect: invalid argument
```

Re-discovered while resuming the parked re-run task from #119 (whose original premise — `waitForSocket` timeout on the sbsh-era socket — is stale after the kuketty phases #413/#415/#417/#420 landed). The socket now appears as expected; the failure has moved from socket-staging to the host-side `connect(2)` call.

**Root cause** (verified). Linux's `sockaddr_un` `sun_path` limit is **108 bytes including the terminating NUL** (`UNIX_PATH_MAX`). When the path passed to `connect()` exceeds that limit, the kernel returns `EINVAL`, which `net.Dial("unix", path)` surfaces verbatim as `connect: invalid argument`. The host-side socket path the e2e test connects to is computed by `fs.ContainerSocketPath` (`internal/util/fs/metadata.go:138`) as:

```
<runPath>/data/<realm>/<space>/<stack>/<cell>/<container>/tty/socket
```

For the e2e scenario, `getRandomRunPath` (`e2e/e2e_test.go:119`) backs `runPath` with `t.TempDir()`, which Go names after the test function: `/tmp/<TestName><rand>/NNN`. `TestKuke_AttachDetach_KeepsTaskRunning` is the longest test name in the e2e suite at 39 characters, and the combined path comes out to **109 bytes** — one over the limit:

```
$ echo -n "/tmp/TestKuke_AttachDetach_KeepsTaskRunning2015614058/001/data/e-r-86/e-sp-52/e-st-4f/e-ce-c4/work/tty/socket" | wc -c
109
```

Inside the container the path is `/run/kukeon/tty/socket` (well under 108) so `kuketty`'s `Listen` works; the same inode is reachable on the host *only* via the long bind-mount source path, and that's where the SUN_PATH overflow bites at `kuke attach` time.

This is a regression hazard beyond the e2e suite: any production runPath whose layout pushes `<runPath>/data/.../tty/socket` over 108 bytes (e.g. operators who place `--run-path` deeper than `/opt/kukeon`, or whose realm/space/stack/cell IDs are longer than the test fixture's `e-r-XX`) will hit the same failure under `kuke attach` with no clear diagnostic.

## Repro

On `origin/main` (6bd245b) with docker + containerd running and `make kuke && make kuketty` already done:

```
$ docker build --build-arg VERSION=v0.0.0-e2e -t kukeon-local:e2e .
$ HOME=$HOME E2E_BIN_DIR=$PWD KUKEON_E2E_IMAGE=docker.io/library/kukeon-local:e2e \
  KUKEON_E2E_IMAGE_DOCKER_NAME=kukeon-local:e2e \
  go test -v -run '^TestKuke_AttachDetach_KeepsTaskRunning$' ./e2e
...
    e2e_kuke_attach_test.go:139: kuke attach exited with code 1 (want 0)
        output:
        Error: error attaching client: ping failed: dial unix /tmp/TestKuke_AttachDetach_KeepsTaskRunning.../001/data/e-r-XX/e-sp-XX/e-st-XX/e-ce-XX/work/tty/socket: connect: invalid argument
--- FAIL: TestKuke_AttachDetach_KeepsTaskRunning (13.79s)
```

`waitForSocket` (`e2e/e2e_kuke_attach_test.go:215`) passes — the per-container socket inode is created — so the failure is purely at host-side `connect()`.

## Proposal

The clean fix is to make the host-side socket path SUN_PATH-safe by construction, instead of letting it inherit the full metadata layout depth. Two options:

1. **Host-side stable shortname.** Maintain a per-container short symlink (e.g. `<runPath>/s/<short-id>` where `<short-id>` is the existing 8-char cellID) that resolves to the same socket inode. `kuke attach` connects via the short symlink; the deep metadata path stays as-is for everything else. The symlink would be staged at the same point the daemon currently materializes the tty directory.
2. **Shorter base layout.** Drop the `/data/<realm>/<space>/<stack>/<cell>/<container>/tty/` segment in favour of `<runPath>/sockets/<cellID>-<container>` for the socket only. The metadata-dir layout stays for files that aren't size-constrained; only `ContainerSocketPath` moves.

Option 1 is the lower-blast-radius call — only `ContainerSocketPath`'s callers on the connect side change, and the symlink target stays the bind-mount source so `kuketty` inside the container is untouched. Option 2 ripples more (every reader of the host-side socket path moves), but it eliminates the indirection.

Either way, the daemon should also bound-check the resolved host-side socket path at provision time and refuse with a clear error if it would exceed 107 bytes (108 minus NUL), so future deepening of the layout fails loud rather than at `kuke attach`.

Finally, separately scope the e2e test: `getRandomRunPath` should fall back to a shorter prefix (e.g. `os.MkdirTemp("/tmp", "ke-")`) when `t.TempDir()` would produce a SUN_PATH-overflowing socket path. This keeps the suite green even before the architectural fix lands, and the harness check protects against test names accidentally drifting longer.

## Acceptance criteria

- [ ] Host-side path returned by `fs.ContainerSocketPath` (or its replacement) for any well-formed `<runPath, realm, space, stack, cell, container>` tuple is `<= 107` bytes for all `runPath` under 60 bytes — verified by a unit test in `internal/util/fs/`.
- [ ] Daemon refuses container provisioning with a clear `socket path %q exceeds SUN_PATH` error when the resolved host-side socket path would overflow, instead of deferring the failure to `kuke attach`.
- [ ] `e2e/e2e_test.go:getRandomRunPath` produces a runPath short enough that `TestKuke_AttachDetach_KeepsTaskRunning` (the longest test name in the suite) is SUN_PATH-safe.
- [ ] `make e2e TestKuke_AttachDetach_KeepsTaskRunning` passes end-to-end on `origin/main` after the fix: socket appears, `kuke attach` connects, Ctrl+] Ctrl+] sequence is consumed by the in-process filter, `kuke attach` exits 0, and the underlying container task remains RUNNING.

## Out of scope

- Lifting the workload `tty/` layout itself — the deep layout is fine for everything that isn't a unix socket; only the socket inode needs a SUN_PATH-safe handle.
- Renaming or merging `internal/util/fs/metadata.go`'s helper set; the constraint is on the resolved socket path, not the structure of the metadata directory.
- Backporting the bound check to merged daemon snapshots that pre-date this issue; the dev-init smoke catches stale-daemon problems on the operator's host.

## References

- #119 — parked re-run of `TestKuke_AttachDetach_KeepsTaskRunning`, closed-as-outdated when this issue lands (per the owner's instruction in https://github.com/eminwux/kukeon/issues/119#issuecomment).
- Kuketty injection-path phases: #413, #415, #417, #420.
- `internal/ctr/attachable.go:54` (`AttachableSocketPath` — the in-container peer of the host path).
- `internal/util/fs/metadata.go:138` (`ContainerSocketPath` — the host-side socket location, single source of truth).
- `e2e/e2e_test.go:119` (`getRandomRunPath` — produces the t.TempDir-rooted runPath that surfaces the overflow).
- Linux `man 7 unix` (`UNIX_PATH_MAX = 108`).

--- a/e2e/e2e_kuke_attach_test.go
+++ b/e2e/e2e_kuke_attach_test.go
@@ -104,7 +104,11 @@ func TestKuke_AttachDetach_KeepsTaskRunning(t *testing.T) {
 	// creates the metadata dir at provision time and sbsh binds the socket
 	// inside the container at task start. If this is missing, pkg/attach
 	// would fail in a way that's hard to attribute, so check explicitly.
-	socketPath := fs.ContainerSocketPath(runPath,
+	// Polls the SUN_PATH-safe symlink the runner stages (issue #521) — the
+	// same handle `kuke attach` connects through. `os.Stat` follows the
+	// symlink, so the loop succeeds once kuketty creates the deep socket
+	// inode at the resolved target.
+	socketPath := fs.ContainerSocketSymlinkPath(runPath,
 		realmName, spaceName, stackName, cellName, "work")
 	waitForSocket(t, socketPath, 10*time.Second)
 

--- a/e2e/e2e_test.go
+++ b/e2e/e2e_test.go
@@ -109,16 +109,52 @@ func runBinary(t *testing.T, env []string, command string, args ...string) (int,
 	return exitCode, []byte(stdoutBuf.String()), []byte(stderrBuf.String())
 }
 
-// getRandomRunPath returns a unique, absolute run path for the test, backed
-// by t.TempDir(). The directory is created up front and removed automatically
-// when the test ends, so callers do not need a separate create-or-cleanup
-// step. Each call returns its own path, which avoids contention on a shared
-// parent directory (issue #491: MkdirAll on a shared parent races
-// concurrent t.Cleanup RemoveAll from sibling parallel tests, surfacing as
-// ENOENT on the new leaf).
+// getRandomRunPath returns a unique, absolute run path for the test. When
+// the t.TempDir-rooted path would push the *deep* per-container kuketty
+// socket path (`<runPath>/data/<realm>/<space>/<stack>/<cell>/<container>/
+// tty/socket`) over Linux's UNIX_PATH_MAX, the helper falls back to a
+// short /tmp prefix and registers a cleanup so the suite stays SUN_PATH-
+// safe regardless of how Go names the t.TempDir parent (issue #521).
+//
+// The architectural fix (issue #521) routes `kuke attach` through a short
+// symlink, so the deep path length no longer breaks `connect(2)` —
+// but the fallback is deliberate defense-in-depth: a future regression
+// that accidentally re-introduces the deep-path dial would fail loudly
+// here instead of in CI.
+//
+// Each call returns its own path so parallel-sibling tests don't race
+// MkdirAll/Cleanup-RemoveAll on a shared parent (issue #491's regression
+// class).
 func getRandomRunPath(t *testing.T) string {
 	t.Helper()
-	return t.TempDir()
+	tempDir := t.TempDir()
+	if deepSocketPathFits(tempDir) {
+		return tempDir
+	}
+	// t.TempDir() is the standard, but here it is exactly what overflows
+	// SUN_PATH because Go derives the temp basename from the test
+	// function name (`TestKuke_AttachDetach_KeepsTaskRunning` is 39
+	// bytes). Drop to a short, name-free /tmp prefix instead.
+	short, err := os.MkdirTemp("/tmp", "ke-") //nolint:usetesting // intentional shorter prefix; see comment
+	if err != nil {
+		t.Fatalf("MkdirTemp(/tmp, ke-): %v", err)
+	}
+	t.Cleanup(func() { _ = os.RemoveAll(short) })
+	return short
+}
+
+// deepSocketPathFits reports whether the longest plausible per-container
+// kuketty socket path under runPath fits SUN_PATH. The longest e2e suffix
+// is `/data/<realm>/<space>/<stack>/<cell>/<container>/tty/socket` with
+// realm/space/stack/cell names produced by generateUnique*Name (`e-r-XX`,
+// `e-sp-XX`, `e-st-XX`, `e-ce-XX`) and the container name fixed to
+// "work" by the attach fixture — 52 bytes of suffix all in. The check is
+// against the deep path, not the SUN_PATH-safe symlink the runner stages,
+// because deep-path regressions are the failure mode this fallback
+// defends against (the symlink budget is comfortably wider).
+func deepSocketPathFits(runPath string) bool {
+	const deepSocketSuffixLen = len("/data/e-r-XX/e-sp-XX/e-st-XX/e-ce-XX/work/tty/socket")
+	return len(runPath)+deepSocketSuffixLen <= consts.KukeonMaxSocketPath
 }
 
 // buildKukeRunPathArgs builds --run-path flag arguments.

--- a/internal/client/local/local.go
+++ b/internal/client/local/local.go
@@ -956,13 +956,20 @@ func controllerImageToWire(img controller.ImageInfo) kukeonv1.ImageInfo {
 // sbsh control-socket path. Bytes never traverse this RPC — the caller
 // (`kuke attach`) opens HostSocketPath directly and runs the sbsh client
 // loop against it.
+//
+// Returns the SUN_PATH-safe symlink path (fs.ContainerSocketSymlinkPath)
+// rather than the deep socket inode path so `connect(2)` never sees a
+// sun_path longer than 107 bytes regardless of how deep the resolved
+// metadata layout is (issue #521). The runner stages the symlink at
+// provision time so a re-derived path on a freshly-spawned `kuke attach`
+// is always pre-materialised.
 func (c *Client) AttachContainer(_ context.Context, doc v1beta1.ContainerDoc) (kukeonv1.AttachContainerResult, error) {
 	spec, err := c.resolveAttachable(doc)
 	if err != nil {
 		return kukeonv1.AttachContainerResult{}, err
 	}
 	return kukeonv1.AttachContainerResult{
-		HostSocketPath: fs.ContainerSocketPath(
+		HostSocketPath: fs.ContainerSocketSymlinkPath(
 			c.ctrl.RunPath(),
 			spec.RealmName, spec.SpaceName, spec.StackName, spec.CellName, spec.ID,
 		),

--- a/internal/consts/consts.go
+++ b/internal/consts/consts.go
@@ -49,6 +49,27 @@ const (
 	// injected by Attachable=true specs.
 	KukeonContainerSocketFile = "socket"
 
+	// KukeonSocketSymlinkSubdir is the basename of the per-RunPath directory
+	// that holds SUN_PATH-safe symlinks to the deep per-container kuketty
+	// sockets (`<RunPath>/data/<realm>/<space>/<stack>/<cell>/<container>/
+	// tty/socket`). Linux `connect(2)` caps the path passed in
+	// sockaddr_un.sun_path at 108 bytes including the terminator (issue #521),
+	// and the deep metadata-rooted path overflows on long RunPaths or
+	// long realm/space/stack/cell IDs. The daemon stages a short symlink at
+	// `<RunPath>/s/<short>` at provision time; `kuke attach` connects via
+	// the symlink so the literal sun_path stays SUN_PATH-safe regardless of
+	// how deep the resolved target lives. Single-letter basename to spend
+	// the budget on the per-container short id, not the directory.
+	KukeonSocketSymlinkSubdir = "s"
+
+	// KukeonMaxSocketPath is Linux's UNIX_PATH_MAX minus the terminating
+	// NUL — the maximum number of bytes a path passed to `connect(2)` on a
+	// unix socket may occupy in sockaddr_un.sun_path. The daemon refuses to
+	// provision an Attachable container whose resolved host-side socket
+	// path would exceed this so the failure surfaces at provision time
+	// rather than at first `kuke attach` (issue #521).
+	KukeonMaxSocketPath = 107
+
 	// KukeonContainerCaptureFile is the basename of the per-container sbsh
 	// capture file inside KukeonContainerTTYDir. sbsh writes the full tty
 	// byte stream — every byte the workload produced and every byte typed

--- a/internal/controller/controller.go
+++ b/internal/controller/controller.go
@@ -224,7 +224,8 @@ func (b *Exec) Close() error {
 // RunPath returns the configured kukeon run path. Surfaced for callers that
 // need to derive host paths from the same root the controller writes to —
 // notably the in-process AttachContainer endpoint, which resolves the
-// per-container sbsh socket via fs.ContainerSocketPath.
+// SUN_PATH-safe per-container kuketty socket via
+// fs.ContainerSocketSymlinkPath.
 func (b *Exec) RunPath() string {
 	return b.opts.RunPath
 }

--- a/internal/controller/runner/attachable.go
+++ b/internal/controller/runner/attachable.go
@@ -25,7 +25,9 @@ import (
 	"os/exec"
 	"path/filepath"
 
+	"github.com/eminwux/kukeon/internal/consts"
 	"github.com/eminwux/kukeon/internal/ctr"
+	"github.com/eminwux/kukeon/internal/errdefs"
 	intmodel "github.com/eminwux/kukeon/internal/modelhub"
 	"github.com/eminwux/kukeon/internal/util/fs"
 	sbshapi "github.com/eminwux/sbsh/pkg/api"
@@ -104,6 +106,10 @@ func attachableTTYDirInitialPerms(kukeonGroupGID int) (os.FileMode, int) {
 func (r *Exec) attachableBuildOpts(_ string, spec intmodel.ContainerSpec, _ []ctr.RegistryCredentials) ([]ctr.BuildOption, error) {
 	if !spec.Attachable {
 		return nil, nil
+	}
+
+	if err := ensureAttachableSocketSymlink(r.opts.RunPath, spec); err != nil {
+		return nil, err
 	}
 
 	ttyDir := fs.ContainerTTYDir(
@@ -553,6 +559,85 @@ func (r *Exec) attachablePostCreateChown(namespace string, spec intmodel.Contain
 		// safety net rather than a hot path.
 	default:
 		return fmt.Errorf("chown %q to (uid=%d, gid=%d): %w", metadataPath, uid, gid, chownErr)
+	}
+	return nil
+}
+
+// attachableSocketSymlinkDirMode is the mode applied to <RunPath>/s when
+// the runner stages it on first Attachable provision. 0o755 mirrors the
+// /opt/kukeon root layout: world-traversable so any process can resolve a
+// staged symlink, world-listable so an operator can `ls` the directory
+// during troubleshooting. The symlinks themselves are never created
+// world-writable (symlink mode is ignored by Linux for symlink-following
+// `connect(2)`; the target inode's mode is what gates access).
+const attachableSocketSymlinkDirMode os.FileMode = 0o0755
+
+// ensureAttachableSocketSymlink stages the SUN_PATH-safe symlink that
+// `kuke attach` connects to (issue #521). The symlink lives at
+// fs.ContainerSocketSymlinkPath under a shallow <RunPath>/s/ subtree and
+// targets the deep fs.ContainerSocketPath inode that kuketty creates inside
+// the bind-mounted /run/kukeon/tty directory at runtime. The target inode
+// does not exist when this function runs — symlinks are pure strings — so
+// the function never blocks on container startup. Recreating the symlink
+// on a re-provision is idempotent: an existing entry is unlinked and
+// rewritten so a prior provision with a stale target gets corrected.
+//
+// The host-side socket path length is the gate: the function refuses to
+// stage a symlink whose resolved path would exceed
+// consts.KukeonMaxSocketPath bytes, surfacing errdefs.ErrSocketPathTooLong
+// with the offending path named. This is the provision-time fail-fast the
+// AC requires so a future operator-configured RunPath that overflows the
+// SUN_PATH budget cannot defer its failure to first `kuke attach`.
+func ensureAttachableSocketSymlink(runPath string, spec intmodel.ContainerSpec) error {
+	symlinkPath := fs.ContainerSocketSymlinkPath(
+		runPath,
+		spec.RealmName, spec.SpaceName, spec.StackName, spec.CellName, spec.ID,
+	)
+	if len(symlinkPath) > consts.KukeonMaxSocketPath {
+		return fmt.Errorf("%w: %q is %d bytes (limit %d)",
+			errdefs.ErrSocketPathTooLong, symlinkPath, len(symlinkPath), consts.KukeonMaxSocketPath)
+	}
+
+	symlinkDir := fs.ContainerSocketSymlinkDir(runPath)
+	if err := os.MkdirAll(symlinkDir, attachableSocketSymlinkDirMode); err != nil {
+		return fmt.Errorf("mkdir %q: %w", symlinkDir, err)
+	}
+
+	target := fs.ContainerSocketPath(
+		runPath,
+		spec.RealmName, spec.SpaceName, spec.StackName, spec.CellName, spec.ID,
+	)
+	// os.Symlink fails with EEXIST when a dirent already lives at the
+	// destination; the re-provision case clears it first so a stale target
+	// from a prior layout does not silently linger. Use os.Remove (not
+	// RemoveAll) so a hostile or buggy actor cannot trick us into deleting a
+	// real directory tree if the destination has somehow turned into one.
+	if err := os.Remove(symlinkPath); err != nil && !errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("remove stale symlink %q: %w", symlinkPath, err)
+	}
+	if err := os.Symlink(target, symlinkPath); err != nil {
+		return fmt.Errorf("symlink %q -> %q: %w", symlinkPath, target, err)
+	}
+	return nil
+}
+
+// removeAttachableSocketSymlink unlinks the SUN_PATH-safe symlink staged by
+// ensureAttachableSocketSymlink. Best-effort: a missing entry is not an
+// error (idempotent under repeated cell / container deletes), but any
+// other failure is returned so the caller can surface the operator-
+// actionable case (e.g. permissions) without losing the signal. Skips
+// non-Attachable specs so the cell-delete path can call it
+// unconditionally.
+func removeAttachableSocketSymlink(runPath string, spec intmodel.ContainerSpec) error {
+	if !spec.Attachable {
+		return nil
+	}
+	symlinkPath := fs.ContainerSocketSymlinkPath(
+		runPath,
+		spec.RealmName, spec.SpaceName, spec.StackName, spec.CellName, spec.ID,
+	)
+	if err := os.Remove(symlinkPath); err != nil && !errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("remove socket symlink %q: %w", symlinkPath, err)
 	}
 	return nil
 }

--- a/internal/controller/runner/delete_cell.go
+++ b/internal/controller/runner/delete_cell.go
@@ -105,6 +105,22 @@ func (r *Exec) DeleteCell(cell intmodel.Cell) error {
 
 	// Workload containers tracked on the cell spec.
 	for _, containerSpec := range internalCell.Spec.Containers {
+		// Unlink the per-container SUN_PATH-safe socket symlink staged at
+		// provision time (issue #521) before tearing down the container.
+		// The symlink lives at <RunPath>/s/<short>, outside the
+		// CellMetadataDir tree that the RemoveAll below would otherwise
+		// reach — leaving it behind would accumulate stale symlinks
+		// pointing at vanished bind-mount sources across re-creates. The
+		// helper short-circuits on non-Attachable specs.
+		if symlinkErr := removeAttachableSocketSymlink(r.opts.RunPath, containerSpec); symlinkErr != nil {
+			r.logger.WarnContext(
+				r.ctx,
+				"failed to remove socket symlink",
+				"container", containerSpec.ID,
+				"error", symlinkErr,
+			)
+		}
+
 		containerID := containerSpec.ContainerdID
 		if containerID == "" {
 			// A partial init can persist the cell document before the

--- a/internal/controller/runner/delete_container.go
+++ b/internal/controller/runner/delete_container.go
@@ -194,5 +194,18 @@ func (r *Exec) DeleteContainer(cell intmodel.Cell, containerID string) error {
 		_ = r.purgeCNIForContainer(containerdID, netnsPath, networkName)
 	}
 
+	// Unlink the SUN_PATH-safe socket symlink staged at provision time
+	// (issue #521). The symlink lives outside the cell metadata tree, so
+	// the CellMetadataDir RemoveAll path that delete_cell.go runs doesn't
+	// reach it; a missing dirent is fine — the helper is idempotent.
+	if symlinkErr := removeAttachableSocketSymlink(r.opts.RunPath, *foundContainerSpec); symlinkErr != nil {
+		r.logger.WarnContext(
+			r.ctx,
+			"failed to remove socket symlink",
+			"container", containerID,
+			"error", symlinkErr,
+		)
+	}
+
 	return nil
 }

--- a/internal/errdefs/errdefs.go
+++ b/internal/errdefs/errdefs.go
@@ -224,6 +224,16 @@ var (
 	// non-root attachable containers.
 	ErrAttachNoCandidate = errors.New("no attachable container in cell")
 
+	// ErrSocketPathTooLong fires when the resolved host-side path of a
+	// per-container kuketty control socket would overflow Linux's
+	// sockaddr_un.sun_path buffer (consts.KukeonMaxSocketPath bytes plus
+	// the terminating NUL — UNIX_PATH_MAX = 108). The daemon refuses to
+	// provision the Attachable container so the failure surfaces at
+	// provision time with the offending path named, instead of deferring
+	// the failure to `kuke attach` where the kernel-level `connect(2)`
+	// returns the cryptic `invalid argument` (issue #521).
+	ErrSocketPathTooLong = errors.New("host-side kuketty socket path exceeds SUN_PATH")
+
 	// Profile-related errors.
 
 	// ErrProfileNotFound is returned when `kuke run -p <name>` cannot find

--- a/internal/util/fs/metadata.go
+++ b/internal/util/fs/metadata.go
@@ -182,15 +182,18 @@ func ContainerSocketSymlinkPath(baseRunPath, realmName, spaceName, stackName, ce
 
 // containerSocketShortID returns the 16-hex digest that names the per-
 // container symlink inside ContainerSocketSymlinkDir. Derived from
-// sha256("<realm>|<space>|<stack>|<cell>|<container>")[:8] so the same
-// identity tuple always resolves to the same name across processes; the
-// pipe separator avoids any concatenation collision between values that
-// happen to share a suffix/prefix (e.g. realm "ab" + space "c" vs realm
-// "a" + space "bc"). 64 bits of randomness puts collisions astronomically
-// out of reach for any plausible per-host container count.
+// sha256("<realm>_<space>_<stack>_<cell>_<container>")[:8] so the same
+// identity tuple always resolves to the same name across processes. The
+// "_" separator is chosen because naming.ValidateRealmName and
+// naming.ValidateHierarchyName both reject "_" (and "/") in any
+// identifier — every other byte, including "|", is a legal name
+// character — so the join is unambiguous: two distinct tuples cannot
+// produce the same concatenation regardless of any byte values inside
+// the names. 64 bits of randomness puts collisions astronomically out
+// of reach for any plausible per-host container count.
 func containerSocketShortID(realmName, spaceName, stackName, cellName, containerName string) string {
 	digest := sha256.Sum256([]byte(
-		realmName + "|" + spaceName + "|" + stackName + "|" + cellName + "|" + containerName,
+		realmName + "_" + spaceName + "_" + stackName + "_" + cellName + "_" + containerName,
 	))
 	return hex.EncodeToString(digest[:8])
 }

--- a/internal/util/fs/metadata.go
+++ b/internal/util/fs/metadata.go
@@ -17,6 +17,8 @@
 package fs
 
 import (
+	"crypto/sha256"
+	"encoding/hex"
 	"encoding/json"
 	"fmt"
 	"path/filepath"
@@ -131,15 +133,66 @@ func ContainerTTYDir(baseRunPath, realmName, spaceName, stackName, cellName, con
 	)
 }
 
-// ContainerSocketPath returns the host-side path for a container's sbsh
-// terminal socket. This is the single source of truth for both the
-// host-visible path that `kuke attach` connects to and the path sbsh listens
-// on inside the container (mediated by the ContainerTTYDir bind mount).
+// ContainerSocketPath returns the host-side path of the kuketty terminal
+// socket inode for a container — the bind-mount source path the kuketty
+// listener inside the container resolves to via /run/kukeon/tty/socket.
+// This path can be arbitrarily long (it inherits the full metadata layout
+// depth) and is NOT safe to pass to `connect(2)`; see
+// ContainerSocketSymlinkPath for the SUN_PATH-safe handle that
+// `kuke attach` uses.
 func ContainerSocketPath(baseRunPath, realmName, spaceName, stackName, cellName, containerName string) string {
 	return filepath.Join(
 		ContainerTTYDir(baseRunPath, realmName, spaceName, stackName, cellName, containerName),
 		consts.KukeonContainerSocketFile,
 	)
+}
+
+// ContainerSocketSymlinkDir returns the per-RunPath directory that holds
+// SUN_PATH-safe symlinks to per-container kuketty sockets. The daemon
+// creates this directory once and stages one symlink per Attachable
+// container inside it; cell / container teardown removes the symlinks but
+// leaves the directory itself in place (idempotent — re-provisioning the
+// same container regenerates its symlink).
+func ContainerSocketSymlinkDir(baseRunPath string) string {
+	return filepath.Join(baseRunPath, consts.KukeonSocketSymlinkSubdir)
+}
+
+// ContainerSocketSymlinkPath returns the SUN_PATH-safe host-side path
+// `kuke attach` connects to. It is a short symlink staged by the daemon
+// inside ContainerSocketSymlinkDir that resolves to the deep socket inode
+// returned by ContainerSocketPath. Linux's sockaddr_un.sun_path is bounded
+// at 108 bytes including the terminator (consts.KukeonMaxSocketPath = 107
+// usable bytes); the deep metadata-rooted ContainerSocketPath overflows on
+// long RunPaths or long realm/space/stack/cell IDs (issue #521), so
+// every host-side `connect(2)` is routed through this short handle.
+//
+// The basename is a 16-hex (sha256[:8]) digest of the
+// realm|space|stack|cell|container tuple, which is deterministic across
+// process restarts (so a freshly-spawned `kuke attach` re-derives the same
+// path the provisioning runner staged) and uniquely identifies a
+// container without inheriting the depth of the metadata layout. With a
+// RunPath ≤ 60 bytes the resolved path is `<RunPath>/s/<16hex>` ≤ 79
+// bytes — well inside SUN_PATH.
+func ContainerSocketSymlinkPath(baseRunPath, realmName, spaceName, stackName, cellName, containerName string) string {
+	return filepath.Join(
+		ContainerSocketSymlinkDir(baseRunPath),
+		containerSocketShortID(realmName, spaceName, stackName, cellName, containerName),
+	)
+}
+
+// containerSocketShortID returns the 16-hex digest that names the per-
+// container symlink inside ContainerSocketSymlinkDir. Derived from
+// sha256("<realm>|<space>|<stack>|<cell>|<container>")[:8] so the same
+// identity tuple always resolves to the same name across processes; the
+// pipe separator avoids any concatenation collision between values that
+// happen to share a suffix/prefix (e.g. realm "ab" + space "c" vs realm
+// "a" + space "bc"). 64 bits of randomness puts collisions astronomically
+// out of reach for any plausible per-host container count.
+func containerSocketShortID(realmName, spaceName, stackName, cellName, containerName string) string {
+	digest := sha256.Sum256([]byte(
+		realmName + "|" + spaceName + "|" + stackName + "|" + cellName + "|" + containerName,
+	))
+	return hex.EncodeToString(digest[:8])
 }
 
 // ContainerCapturePath returns the host-side path of the per-container sbsh

--- a/internal/util/fs/metadata_test.go
+++ b/internal/util/fs/metadata_test.go
@@ -93,8 +93,15 @@ func TestContainerSocketSymlinkPath_DistinctPerContainer(t *testing.T) {
 	// Two distinct identity tuples must map to distinct short ids; sharing
 	// would collapse two containers onto the same dirent and the second
 	// provision would silently rewrite the first's target. Vary exactly
-	// one position per row so a coverage gap surfaces clearly. The pipe-
-	// separator concat-collision pair ("ab|c" vs "a|bc") is the last row.
+	// one position per row so a coverage gap surfaces clearly.
+	//
+	// The "pipe-in-name" row is the regression guard for the separator
+	// choice itself: "|" is a legal name byte (only "_" and "/" are
+	// rejected by naming.ValidateRealmName / naming.ValidateHierarchyName),
+	// so any non-"_"/"/" separator would let realm "a|b" + space "c"
+	// collide with realm "a" + space "b|c". The "concat" row is the
+	// classic split-shift case ("ab"+"c" vs "a"+"bc") that any
+	// fixed-byte separator catches.
 	type pair struct {
 		name string
 		a, b [5]string
@@ -106,6 +113,7 @@ func TestContainerSocketSymlinkPath_DistinctPerContainer(t *testing.T) {
 		{"space", [5]string{"r", "sA", "k", "c", "c1"}, [5]string{"r", "sB", "k", "c", "c1"}},
 		{"realm", [5]string{"rA", "s", "k", "c", "c1"}, [5]string{"rB", "s", "k", "c", "c1"}},
 		{"concat", [5]string{"ab", "c", "k", "c", "c1"}, [5]string{"a", "bc", "k", "c", "c1"}},
+		{"pipe-in-name", [5]string{"a|b", "c", "k", "c", "c1"}, [5]string{"a", "b|c", "k", "c", "c1"}},
 	}
 	for _, p := range pairs {
 		a := fs.ContainerSocketSymlinkPath("/opt/kukeon", p.a[0], p.a[1], p.a[2], p.a[3], p.a[4])

--- a/internal/util/fs/metadata_test.go
+++ b/internal/util/fs/metadata_test.go
@@ -18,6 +18,7 @@ package fs_test
 
 import (
 	"encoding/json"
+	"strings"
 	"testing"
 
 	"github.com/eminwux/kukeon/internal/apischeme"
@@ -60,6 +61,74 @@ func TestContainerSocketPath(t *testing.T) {
 		consts.KukeonContainerTTYDir + "/" + consts.KukeonContainerSocketFile
 	if got != want {
 		t.Errorf("ContainerSocketPath = %q, want %q", got, want)
+	}
+}
+
+func TestContainerSocketSymlinkPath_Layout(t *testing.T) {
+	// The SUN_PATH-safe symlink lives under <RunPath>/s/<short> — a single-
+	// letter parent so the budget goes to the per-container short id, not
+	// the directory. Anchored here so a refactor that drifts the layout
+	// (e.g. promotes the symlink dir to a per-realm subtree) breaks the
+	// test instead of silently violating the SUN_PATH bound.
+	got := fs.ContainerSocketSymlinkPath("/opt/kukeon", "realm", "space", "stack", "cell", "c1")
+	wantPrefix := "/opt/kukeon/" + consts.KukeonSocketSymlinkSubdir + "/"
+	if !strings.HasPrefix(got, wantPrefix) {
+		t.Errorf("symlink path = %q, want prefix %q", got, wantPrefix)
+	}
+}
+
+func TestContainerSocketSymlinkPath_Deterministic(t *testing.T) {
+	// The short id is sha256[:8]-derived, so the same tuple must round-trip
+	// to the same path across calls. `kuke attach` re-derives the path from
+	// the container doc rather than reading the symlink dirent, so a
+	// process restart that changed the hash would silently break attach.
+	a := fs.ContainerSocketSymlinkPath("/opt/kukeon", "realm", "space", "stack", "cell", "c1")
+	b := fs.ContainerSocketSymlinkPath("/opt/kukeon", "realm", "space", "stack", "cell", "c1")
+	if a != b {
+		t.Errorf("symlink path is not deterministic: a=%q b=%q", a, b)
+	}
+}
+
+func TestContainerSocketSymlinkPath_DistinctPerContainer(t *testing.T) {
+	// Two distinct identity tuples must map to distinct short ids; sharing
+	// would collapse two containers onto the same dirent and the second
+	// provision would silently rewrite the first's target. Vary exactly
+	// one position per row so a coverage gap surfaces clearly. The pipe-
+	// separator concat-collision pair ("ab|c" vs "a|bc") is the last row.
+	type pair struct {
+		name string
+		a, b [5]string
+	}
+	pairs := []pair{
+		{"container", [5]string{"r", "s", "k", "c", "c1"}, [5]string{"r", "s", "k", "c", "c2"}},
+		{"cell", [5]string{"r", "s", "k", "cA", "c1"}, [5]string{"r", "s", "k", "cB", "c1"}},
+		{"stack", [5]string{"r", "s", "kA", "c", "c1"}, [5]string{"r", "s", "kB", "c", "c1"}},
+		{"space", [5]string{"r", "sA", "k", "c", "c1"}, [5]string{"r", "sB", "k", "c", "c1"}},
+		{"realm", [5]string{"rA", "s", "k", "c", "c1"}, [5]string{"rB", "s", "k", "c", "c1"}},
+		{"concat", [5]string{"ab", "c", "k", "c", "c1"}, [5]string{"a", "bc", "k", "c", "c1"}},
+	}
+	for _, p := range pairs {
+		a := fs.ContainerSocketSymlinkPath("/opt/kukeon", p.a[0], p.a[1], p.a[2], p.a[3], p.a[4])
+		b := fs.ContainerSocketSymlinkPath("/opt/kukeon", p.b[0], p.b[1], p.b[2], p.b[3], p.b[4])
+		if a == b {
+			t.Errorf("[%s] distinct tuples collided: a=%q b=%q", p.name, a, b)
+		}
+	}
+}
+
+func TestContainerSocketSymlinkPath_FitsSUNPath(t *testing.T) {
+	// AC #1 of issue #521: for any well-formed tuple under a RunPath
+	// ≤ 60 bytes, the resolved host-side socket path must stay under
+	// consts.KukeonMaxSocketPath so `connect(2)` never trips SUN_PATH
+	// overflow at `kuke attach` time. The symlink path budget is
+	// independent of how deep the metadata layout becomes — that is the
+	// whole point of routing the dial through it.
+	longRunPath := "/" + strings.Repeat("x", 59) // 60 bytes total
+	longName := strings.Repeat("z", 64)          // operator-supplied names can be long
+	got := fs.ContainerSocketSymlinkPath(longRunPath, longName, longName, longName, longName, longName)
+	if len(got) > consts.KukeonMaxSocketPath {
+		t.Errorf("symlink path overflows SUN_PATH: %d > %d (path=%q)",
+			len(got), consts.KukeonMaxSocketPath, got)
 	}
 }
 

